doc/internals/export.rst

```diff
@@ -0,0 +1,98 @@
+.. _export:
+
+Exporting Symbols
+=================
+
+The arbor libraries are compiled with hidden visibility by default. Thus, symbols which are part of
+the public API need to be marked explicitely as visible. Arbor provides a couple of macros to
+annotate functions and classes which are defined in the header file ``export.hpp`` in each library's
+include directory, i.e. ``include/arbor/export.hpp``. These header files are generated at configure
+time based on the build variant, compiler and platform.
+
+Macro Descripiton
+-----------------
+
+.. c:macro:: ARB_LIBNAME_API
+
+    Here ``LIBNAME`` is a placeholder for the library's name: ``ARB_ARBOR_API`` for the main arbor
+    library, ``ARB_ARBORIO_API`` for arborio, etc. This macro is intended to annotate functions,
+    classes and structs which need to be accessible when using the library. Note, that it expands to
+    different values when arbor is being built vs. when arbor is being used by an application. Below
+    we list the places where the macro needs to be added or can be safely omitted (we assume all of
+    the symbols below are part of the public API).
+
+    .. code-block:: cpp
+        :caption: header.hpp
+
+        // free function declaration
+        ARB_ARBOR_API void foo();
+
+        // free function (inline)
```

Comment:
'inline' refers to not being exposed/hidden? If so, prefer 'hidden'. If not, educate me :)

Reply:
inline means, the function body (the definition) of the function resides in the header (it is compiled anew for each usage of Arbor). Conversely non-inline functions are compiled once and are part of the pre-compiled library (only the declaration is in the header).

Reply:
Of course, the absence of `ARB_ARBOR_API` ticked me off.

Then there are two differences with the 'free function': absence of the annotation and the presence of a body. Whether a free function is inline or not, does not affect how to annotate, right? If so, maybe leave it out.

Reply:
it does affect annotation:
- inline free function (definition in header): don't annotate (is visible no matter what)
- free function (declaration in header, definition in source): must annotate both declaration and definition (is hidden otherwise)

Reply:
Can you put those bullets in the documentation? Very clear :)

Reply:
I did add 2 bullets in the text - let me know if that's ok

Reply:
Yes!